src/tests/payment/test_stripe_webhook.py

```diff
@@ -0,0 +1,371 @@
+import hashlib
+import hmac
+import json
+import time
+from datetime import datetime, timedelta
+from decimal import Decimal
+from unittest import mock
+
+import pytest
+from django.test import Client
+from django.urls import reverse_lazy
+from django.utils import timezone
+
+from payment.models import CustomerDetail, Subscription, User
+
+PASSWORD = "password1"
+FIRSTNAME = "John"
+LASTNAME = "Doe"
+EMAIL = "johndoe@example.fr"
+USERNAME = EMAIL
+EMAIL_HOST_USER = "support@ordopro.fr"
+
+STRIPE_WEBHOOK_SECRET = "whsec_testsecret"
+
+
+@pytest.fixture
+def user(db):
+    """Creates and yields a new user."""
+    user = User.objects.create(
+        username=USERNAME, first_name=FIRSTNAME, last_name=LASTNAME, email=EMAIL
+    )
+    user.set_password(PASSWORD)
+    user.save()
+    yield user
+    user.delete()
+
+
+@pytest.fixture
+def client():
+    return Client()
+
+
+@pytest.fixture
+def payload():
+    """Fixture to generate a test payload."""
+    return {
+        "type": "checkout.session.completed",
+        "data": {
+            "object": {
+                "id": "cs_test_a1SdT2v81sae3cW",
+                "object": "checkout.session",
+                "amount_subtotal": 990,
+                "amount_total": 990,
+                "billing_address_collection": "required",
+                "cancel_url": "http://127.0.0.1:8000/api/v1/subscriptions/cancel/",
+                "created": 1731848739,
+                "currency": "eur",
+                "custom_fields": [],
+                "customer": "cus_REbNQXKKFCRF2c",
+                "customer_creation": "always",
+                "customer_details": {
+                    "address": {
+                        "city": "Saint-Ouen-l'Aumône",
+                        "country": "FR",
+                        "line1": "1 Rue Pagnere",
+                        "line2": "null",
+                        "postal_code": "95310",
+                        "state": "null",
+                    },
+                    "email": "kadi@test.com",
+                    "name": "kadi",
+                },
+                "customer_email": "kadi@test.com",
+                "expires_at": 1731935139,
+                "metadata": {"product_name": "Essentiel", "user_id": "5"},
+                "mode": "subscription",
+                "payment_status": "paid",
+                "status": "complete",
+            },
+        },
+    }
+
+
+def generate_stripe_signature(payload, secret):
+    """
+    Simulates the Stripe signature for the test using the secret key.
+    """
+    payload_str = json.dumps(payload)
+    timestamp = str(int(time.time()))
+    sig_base = f"{timestamp}.{payload_str}"
+    signature = hmac.new(
+        secret.encode("utf-8"), sig_base.encode("utf-8"), hashlib.sha256
+    ).hexdigest()
+    return f"t={timestamp},v1={signature}"
```

Comment:
Can't we use one of the stripe built-in functions rather than rewriting it ourselves?

Reply:
Stripe does not provide a native method to generate a webhook signature. Therefore, we need to stick with the current method, which uses `hmac` to simulate the signature generation, as already done in our `generate_stripe_signature` function. This remains the best approach for testing and simulating webhook signatures for Stripe.